examples/gaussian_process/plot_gpr_noisy.py

```diff
@@ -3,95 +3,183 @@
-plt.title("Initial: %s\nOptimum: %s\nLog-Marginal-Likelihood: %s"
-          % (kernel, gp.kernel_,
-             gp.log_marginal_likelihood(gp.kernel_.theta)))
-plt.tight_layout()
-
-# Plot LML landscape
-plt.figure()
-theta0 = np.logspace(-2, 3, 49)
-theta1 = np.logspace(-2, 0, 50)
-Theta0, Theta1 = np.meshgrid(theta0, theta1)
-LML = [[gp.log_marginal_likelihood(np.log([0.36, Theta0[i, j], Theta1[i, j]]))
-        for i in range(Theta0.shape[0])] for j in range(Theta0.shape[1])]
-LML = np.array(LML).T
-
-vmin, vmax = (-LML).min(), (-LML).max()
-vmax = 50
-level = np.around(np.logspace(np.log10(vmin), np.log10(vmax), 50), decimals=1)
-plt.contour(Theta0, Theta1, -LML,
-            levels=level, norm=LogNorm(vmin=vmin, vmax=vmax))
+# %%
+plt.plot(X, y, label="Perfect generator")
+plt.scatter(x=X_train[:, 0], y=y_train, color="tab:orange", label="Noisy measurement")
+plt.errorbar(X, y_mean, y_std)
+plt.legend()
+plt.xlabel("X")
+plt.ylabel("y")
+_ = plt.title(
+    f"Initial: {kernel}\nOptimum: {gpr.kernel_}\nLog-Marginal-Likelihood: "
+    f"{gpr.log_marginal_likelihood(gpr.kernel_.theta)}",
+    fontsize=8,
+)
+# %%
+# We see that the optimum kernel found still have a high noise level and
+# an even larger length scale. Furthermore, we observe that the
+# model does not provide faithful predictions.
+#
+# Now, we will initialize the
+# :class:`~sklearn.gaussian_process.kernels.RBF` with a
+# larger length_scale and the
+# :class:`~sklearn.gaussian_process.kernels.WhiteKernel`
+# with a smaller noise level lower bound.
+kernel = 1.0 * RBF(length_scale=1e-1, length_scale_bounds=(1e-2, 1e3)) + WhiteKernel(
+    noise_level=1e-2, noise_level_bounds=(1e-10, 1e1)
+)
+gpr = GaussianProcessRegressor(kernel=kernel, alpha=0.0)
+gpr.fit(X_train, y_train)
+y_mean, y_std = gpr.predict(X, return_std=True)
+
+# %%
+plt.plot(X, y, label="Perfect generator")
+plt.scatter(x=X_train[:, 0], y=y_train, color="tab:orange", label="Noisy measurement")
+plt.errorbar(X, y_mean, y_std)
+plt.legend()
+plt.xlabel("X")
+plt.ylabel("y")
+_ = plt.title(
+    f"Initial: {kernel}\nOptimum: {gpr.kernel_}\nLog-Marginal-Likelihood: "
+    f"{gpr.log_marginal_likelihood(gpr.kernel_.theta)}",
+    fontsize=8,
+)
+
+# %%
+# First, we see that the model's predictions are precise than the previous
+# model's. This new model is able to estimate the noise-free functional
+# relationship.
+#
+# Looking at the kernel hyperparameters, we see that the best combination found
+# has a smaller noise level and shorter length scale than the first model.
+#
+# We can have a look at the Log-Marginal-Likelihood (LML) of GPR for different
+# hyperparameters to get a sense of the local minima.
+from matplotlib.colors import LogNorm
+
+length_scale = np.logspace(-2, 4, num=50)
+noise_level = np.logspace(-2, 1, num=50)
+length_scale_grid, noise_level_grid = np.meshgrid(length_scale, noise_level)
+
+log_marginal_likelihood = [
+    gpr.log_marginal_likelihood(np.log([0.36, scale, noise]))
```

Comment:
I think inspecting the LML is insightful. :+1: 

I don't know enough about GP, but wouldn't the LML depend on the fitted instance of `GaussianProcessRegressor` and its kernel here?

Reply:
Indeed, this is a method. I should add

```suggestion
    gpr.log_marginal_likelihood(theta=np.log([0.36, scale, noise]))
```

If `theta=None`, it will use the fitted parameter of the kernel that maximizes the LML. However, if you pass some parameter, it will recompute the LML for these parameters.